File: AI LAB/02_UninformedSearchTech/BFS.java
package Uninformed_Search_Techniques;
import java.util.*;

public class BFS {
    public static void bfs(int start,int[] visit, ArrayList<ArrayList<Integer>> ans ){
    Queue<Integer> q=new LinkedList<>();
    q.add(start);
    visit[start]=1;

    while (q.size()>0) {
        int front=q.remove();
        System.out.print(front+" ");
        for(int ele:ans.get(front)){
            if(visit[ele]!=1 ){
                q.add(ele);
                visit[ele]=1;
            }
        }
    }
    }

    public static void addEdge(int start,int dest ,ArrayList<ArrayList<Integer>> ans){
        ans.get(start).add(dest);
    }

    public static void main(String[] args) {
        int n=6;
        ArrayList<ArrayList<Integer>> ans=new ArrayList<>();
        
        for(int i=0;i<n;i++){
            ans.add(new ArrayList<>());
        }

        int[] visit=new int[n];

        addEdge(0,5,ans);
        addEdge(0,1,ans);
        addEdge(0,4,ans);
        addEdge(1,0,ans);
        addEdge(1,4,ans);
        addEdge(1,2,ans);
        addEdge(2,1,ans);
        addEdge(2,3,ans);
        addEdge(3,2,ans);
        addEdge(3,4,ans);
        addEdge(4,0,ans);
        addEdge(4,1,ans);
        addEdge(4,3,ans);
        addEdge(0,5,ans);

        


        bfs(1,visit,ans);

        
    }
    
}

File: AI LAB/02_UninformedSearchTech/BFS.cpp
// here the d= depth of shallowest solution and b is a node at every state.
// T (b) = 1+b2+b3+.......+ bd= O (b^d);
// S.C = O(b^d);


#include <bits/stdc++.h>
using namespace std;

void bfs(int start, vector<int> &visit, vector<vector<int>> &graph) {
    queue<int> q;
    q.push(start);
    visit[start] = 1;

    while (!q.empty()) {
        int front = q.front();
        q.pop();
        cout << front << " ";
        for (int ele : graph[front]) {
            if (visit[ele] != 1) { 
                q.push(ele);
                visit[ele] = 1;
            }
        }
    }
}

void addEdge(int start, int dest, vector<vector<int>> &graph) {
    graph[start].push_back(dest);
}

int main() {
    int n = 6;
    vector<vector<int>> graph(n);
    vector<int> visit(n, 0);

    // Adding edges (Adjust indices to be 0-based)
    addEdge(0, 1, graph);
    addEdge(0, 4, graph);
    addEdge(0, 5, graph);
    addEdge(1, 0, graph);
    addEdge(1, 2, graph);
    addEdge(1, 4, graph);
    addEdge(2, 3, graph);
    addEdge(3, 2, graph);
    addEdge(3, 4, graph);
    addEdge(4, 0, graph);
    addEdge(4, 1, graph);
    addEdge(4, 3, graph);

    cout << "The BFS is: ";
    bfs(0, visit, graph);

    return 0;
}
